[Xtensa] Implement lowering Mul/Div/Shift operations.

andreisfr opened the pull request:
Implement lowering of the Mul/Div operations and also shift parts operations. Implement lowering of the bit manipulations, like ROT/SWAP/CTPOP/CTTZ/CTLZ.

andreisfr committed:
[Xtensa] Implement lowering Mul/Div/Shift/ROT/CTTZ/CTLZ/CTPOP operations.

Implement lowering of the Mul/Div operations and also shift
parts operations. Implement lowering of the bit manipulations,
like ROT/SWAP/CTPOP/CTTZ/CTLZ.

andreisfr commented:
@arsenm , @jyknight, @MaskRay , @s-barannikov , PTAL to this PR whenever you have the time.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -665,12 +693,30 @@ SDValue XtensaTargetLowering::getAddrPCRel(SDValue Op,
 SDValue XtensaTargetLowering::LowerConstantPool(ConstantPoolSDNode *CP,
                                                 SelectionDAG &DAG) const {
   EVT PtrVT = getPointerTy(DAG.getDataLayout());
+  auto C = const_cast<Constant *>(CP->getConstVal());
+  auto T = const_cast<Type *>(CP->getType());
```
Shouldn't need const_cast?

andreisfr replied:
Sorry,  if I understand correctly, I removed const_cast

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -665,12 +693,30 @@ SDValue XtensaTargetLowering::getAddrPCRel(SDValue Op,
 SDValue XtensaTargetLowering::LowerConstantPool(ConstantPoolSDNode *CP,
                                                 SelectionDAG &DAG) const {
   EVT PtrVT = getPointerTy(DAG.getDataLayout());
```
Always reuse the type of the value you always have, you should practically never need to call getPointerTy

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -665,12 +693,30 @@ SDValue XtensaTargetLowering::getAddrPCRel(SDValue Op,
+    auto GV = new GlobalVariable(
+        *M, T, /*isConstant=*/true, GlobalVariable::InternalLinkage, C,
+        Twine(DAG.getDataLayout().getPrivateGlobalPrefix()) + "CP" +
+            Twine(DAG.getMachineFunction().getFunctionNumber()) + "_" +
```
DAG.getMachineFunction() is repeated many times, use an MF variable

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -713,6 +759,131 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+  return DAG.getMergeValues(Ops, DL);
+}
+
+SDValue XtensaTargetLowering::LowerMUL(SDValue Op, SelectionDAG &DAG) const {
```
This looks like a simple optimization combine that doesn't belong as custom lowering, and should already be done by the generic combiner

andreisfr replied:
It seems MUL is expanded to "_mulsi3" call instead off "shl" operations for such cases. And stack operations in such case contain many "_mulsi3" calls, instead of just use simple ADDX2(Add with Shift by 1)/ADDX4/ADDX8 Xtensa operations which are generated from (add (shl)) pattern

s-barannikov replied:
You need to override `decomposeMulByConstant` for this to happen.

andreisfr replied:
Thank you for advice, I implemented decomposeMulByConstant

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -665,12 +693,30 @@ SDValue XtensaTargetLowering::getAddrPCRel(SDValue Op,
+                                         CP->getAlign(), CP->getOffset());
+    } else {
+      report_fatal_error("This constantpool type is not supported yet");
+    }
```
Missing constant pool tests for all of these cases, particularly the aggregate and vector handling

andreisfr replied:
Thank you for comments, I added test. Also I removed "Vector" case handling, probably it is early, I will implement it in future.

andreisfr committed:
[Xtensa] Minor fixes in constant pool lowering.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -662,15 +690,33 @@ SDValue XtensaTargetLowering::getAddrPCRel(SDValue Op,
+  if (T->isAggregateType()) {
+    MachineFunction &MF = DAG.getMachineFunction();
+    auto AFI = MF.getInfo<XtensaFunctionInfo>();
+    auto M = const_cast<Module *>(MF.getFunction().getParent());
```
Should not const cast

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -662,15 +690,33 @@ SDValue XtensaTargetLowering::getAddrPCRel(SDValue Op,
+    MachineFunction &MF = DAG.getMachineFunction();
+    auto AFI = MF.getInfo<XtensaFunctionInfo>();
+    auto M = const_cast<Module *>(MF.getFunction().getParent());
+    auto GV = new GlobalVariable(
```
Should not be mutating the IR during selection

arsenm replied:
ping

andreisfr replied:
@arsenm , I'm sorry, you suggest to not use creation of GobalVariable in this case?

arsenm replied:
Yes

andreisfr replied:
Fixed, I implemented CTPOP custom lowering and  removed Global Variable creation.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -827,9 +1010,70 @@ XtensaTargetLowering::emitSelectCC(MachineInstr &MI,
   switch (MI.getOpcode()) {
   case Xtensa::SELECT:
     return emitSelectCC(MI, MBB);
+  case Xtensa::SHL_P: {
```
Instruction selection should be done in `XtensaDAGToDAGISel::Select`.

andreisfr replied:
I added "usesCustomInserter" attribute to the  SHL_P and some other shift pseudo operations. Should I remove this operations processing from EmitInstrWithCustomInserter or maybe remove "usesCustomInserter" attribute?

s-barannikov replied:
This code should be moved to XtensaISelDAGToDAG.cpp and SHL_P and others removed.

andreisfr replied:
fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -98,6 +100,32 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+  setOperationAction(ISD::SDIVREM, MVT::i32, Expand);
+  setOperationAction(ISD::UDIVREM, MVT::i32, Expand);
+
+  setOperationAction(ISD::SHL_PARTS, MVT::i32, Custom);
```
I would expect ordinary 32-bit shifts supported first.

andreisfr replied:
Sorry, maybe I misunderstand something. I implemented patterns for SHL/SHR/SRA operations with register and immediate operands and tests and we can generate 32-bit Xtensa shift instructions now.

andreisfr committed:
[Xtensa] Minor code formatting.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.h:
```diff
@@ -40,6 +40,10 @@ enum {
   SELECT_CC,
+
+  // Shift
+  SRCL,
```
How are these different from ISD opcodes? They deserve a comment.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -662,15 +690,33 @@ SDValue XtensaTargetLowering::getAddrPCRel(SDValue Op,
+
+  // Do not use constant pool for aggregate or vector constant types,
+  // in such cases create global variable, for example to store tabel
+  // when we lower CTTZ operation.
```
It may make sense to not use this "table lookup" expansion at all and custom lower CTTZ_ZERO_UNDEF to a libcall.
Alternatively, you can custom lower CTPOP to a libcall, in which case CTTZ_ZERO_UNDEF will be expanded as `popcount(~x & (x - 1))`.

andreisfr replied:
I implemented custom CTPOP operation lowering.

andreisfr committed:
[Xtensa] Transform multipy by constant.

Implement decomposeMulByConstant function and remove lowering Mul
operation. Minor fixes in lowering constant pool.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -713,6 +761,118 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+
+bool XtensaTargetLowering::decomposeMulByConstant(LLVMContext &Context, EVT VT,
+                                                  SDValue C) const {
+  if (!VT.isScalarInteger())
+    return false;
+
+  // Omit if data size exceeds.
+  if (VT.getSizeInBits() > 32)
+    return false;
```
Would be good to try to handle the splat vector case

andreisfr replied:
I added splat vector case, PTAL.

andreisfr committed:
[Xtensa] Lower CTPOP operation.

Implement lowering of the CTPOP operation. Also remove global
variable creation code from lowerConstantPool functio. Add vector
case handling to the decomposeMulByConstant function.

s-barannikov commented on llvm/test/CodeGen/Xtensa/ctlz-cttz-ctpop.ll:
```diff
@@ -0,0 +1,566 @@
+declare i32 @llvm.ctpop.i32(i32)
+declare i64 @llvm.ctpop.i64(i64)
+
+define i8 @test_cttz_i8(i8 %a) nounwind {
```
There isn't much sense in testing i8 and i16 because they will be promoted to i32.

andreisfr replied:
fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaMachineFunctionInfo.h:
```diff
@@ -0,0 +1,53 @@
+
+namespace llvm {
+
+class XtensaFunctionInfo : public MachineFunctionInfo {
```
Not related to this PR.

andreisfr committed:
[Xtensa] Move SHL/SRL/SRA/SRC lowering from ISelLowering to ISelDagToDag.
Remove SHL/SRL/SRA/SRC pseudo operations. Remove redundant cttz/ctlz/ctpop tests.
Remove Xtensa MachineFunctionInfo implementation.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelDAGToDAG.cpp:
```diff
@@ -125,12 +125,75 @@ FunctionPass *llvm::createXtensaISelDag(XtensaTargetMachine &TM,
+    SDValue N1 = Node->getOperand(1);
+    auto *C = dyn_cast<ConstantSDNode>(N1);
+    // If C is constant in range [0..15] then we can generate SRLI
+    // instruction using pattern matching, otherwise generate SRL
```
Why don't other shifts need similar treatment?

andreisfr replied:
Fixed

s-barannikov replied:
This was a question :)

s-barannikov replied:
ping, why is it different from the others?

andreisfr replied:
You mean why shift amount range checks for SRL case is different from SRA and SHL?

s-barannikov replied:
Yes

andreisfr replied:
The Xtensa SRLI (shift right by immediate) instruction encoding only support 4-bit shift immediate
"SRLI ar, at, 0..15 "
This immediate case is supported in instruction description
https://github.com/llvm/llvm-project/blob/8cf79c39c4b80efedad0f6eb84e04aca83d16154/llvm/lib/Target/Xtensa/XtensaInstrInfo.td#L160 
So, I created code for SRL case, which checks whether immediate fits to 0..15 range, in other case we should place shift amount in register

The SRAI (shift right arithmetic by immediate) instruction allows  encoding  of the shift immediate in 0..31 range
"SRAI ar, at, 0..31"
https://github.com/llvm/llvm-project/blob/8cf79c39c4b80efedad0f6eb84e04aca83d16154/llvm/lib/Target/Xtensa/XtensaInstrInfo.td#L151

The similar case is for SLLI instruction, it allows  encoding  of the shift immediate in 1..31 range
https://github.com/llvm/llvm-project/blob/8cf79c39c4b80efedad0f6eb84e04aca83d16154/llvm/lib/Target/Xtensa/XtensaInstrInfo.td#L168

s-barannikov replied:
Thanks for the explanation.
There doesn't seem to be a testcase for lshr with amt > 15.

s-barannikov replied:
> There doesn't seem to be a testcase for SRL with amt > 15.

ping

andreisfr replied:
I added test for SRL

s-barannikov commented on llvm/test/CodeGen/Xtensa/bswap.ll:
```diff
@@ -0,0 +1,413 @@
+; RUN: llc -mtriple=xtensa -verify-machineinstrs < %s \
```
All the tests are missing "autogenerated by" annotation.

andreisfr replied:
Fixed

andreisfr committed:
[Xtensa] Minor fixes

andreisfr commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -713,6 +746,171 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+  Hi = DAG.getNode(ISD::SELECT, DL, VT, Cond, HiTrue, HiFalse);
+
+  return DAG.getMergeValues({Lo, Hi}, DL);
+}
```
I corrected code

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelDAGToDAG.cpp:
```diff
@@ -125,12 +125,81 @@ FunctionPass *llvm::createXtensaISelDag(XtensaTargetMachine &TM,
+    SDValue N0 = Node->getOperand(0);
+    SDValue N1 = Node->getOperand(1);
+    auto *C = dyn_cast<ConstantSDNode>(N1);
+    // If C is constant in range [0..31] then we can generate SRAI
```
The constant has to be in that range.

andreisfr replied:
I corrected code and comments for shift operation, PTAL. Hope I understood your comment correctly.

s-barannikov replied:
What I mean is that the constant will always be in the range [1, 31]. Everything else is optimized away by DAG combiner.
That makes the range check redundant. You can change it to assert if you wish.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelDAGToDAG.cpp:
```diff
@@ -125,12 +125,81 @@ FunctionPass *llvm::createXtensaISelDag(XtensaTargetMachine &TM,
+    SDValue N0 = Node->getOperand(0);
+    SDValue N1 = Node->getOperand(1);
+    SDValue N2 = Node->getOperand(2);
+    SDNode *SSR = CurDAG->getMachineNode(Xtensa::SSR, DL, MVT::Glue, N2);
+    SDNode *SRC =
+        CurDAG->getMachineNode(Xtensa::SRC, DL, VT, N0, N1, SDValue(SSR, 0));
```
Seems surprising to me to handle this as a function of selection, and not legalization but it's fine

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -713,6 +746,158 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+    return SDValue();
+
+  // CTPOP expansion:
+  // Val = (Val - (Val >> 1)) & 0x55555555
+  // Val = ((Val >> 2) & 0x33333333) + (Val & 0x33333333)
+  // Val = ((Val >> 4) + Val) & 0x0f0f0f0f
+  // Val = (Val >> 8) + Val
+  // Val = (extract bits [16, 20] from Val) + Val
+  // Val = extract bits [0, 5] from Val
+
```
This looks pretty close to the generic expansion. Can you just pattern match the EXTUIs at the end from the end sequence of the generic expansion in a DAG combine?

s-barannikov replied:
The issue is that if you set CTPOP action to Expand, CTTZ_ZERO_UNDEF expands via table lookup and we get back to GlobalVariable and ConstantPool mess.

I'd just custom lower CTPOP to a libcall to avoid this code duplication (assuming compiling the Linux kernel is not in the near future).

andreisfr replied:
> The issue is that if you set CTPOP action to Expand, CTTZ_ZERO_UNDEF expands via table lookup and we get back to GlobalVariable and ConstantPool mess.
> 
> I'd just custom lower CTPOP to a libcall to avoid this code duplication (assuming compiling the Linux kernel is not in the near future).

You suggestion is to implement call to "__popcountsi2"?

arsenm replied:
It depends on whether you want to use a libcall or not. 

> The issue is that if you set CTPOP action to Expand, CTTZ_ZERO_UNDEF expands via table lookup 

The default expansion logic doesn't always do the right thing but is usually fixable?

arsenm replied:
Custom lowering can also just directly call TLI::expandCTPOP

andreisfr replied:
> Custom lowering can also just directly call TLI::expandCTPOP

I implemented call  to "TLI::expandCTPOP" in lowerCTPOP and improved SRL selection (generate EXTUI for suitable constants) and now CTPOP generic expansion generates the same amount of operation as in my previous implementation.

andreisfr committed:
[Xtensa] Imrpove CTPOP lowering. Minor fixes
int SHL/SRC/SRA operations selection.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelDAGToDAG.cpp:
```diff
@@ -125,12 +125,90 @@ FunctionPass *llvm::createXtensaISelDag(XtensaTargetMachine &TM,
+    SDValue N0 = Node->getOperand(0);
+    SDValue N1 = Node->getOperand(1);
+    auto *C = dyn_cast<ConstantSDNode>(N1);
+    // If C is constant in range [1..31] then we can generate SLLI
```
The constant should always be in this range.

s-barannikov replied:
Please make it consistent with SRA.

andreisfr replied:
Fixed

s-barannikov replied:
The comment still needs fixing.

NB
Sentences in comments should end with a period, here and elsewhere. Sorry for having to remind you.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -827,47 +827,8 @@ SDValue XtensaTargetLowering::LowerShiftRightParts(SDValue Op,
-
-  return DAG.getNode(XtensaISD::EXTUI, DL, VT, Val, DAG.getConstant(0, DL, VT),
-                     DAG.getConstant(6, DL, VT));
+  auto &TLI = DAG.getTargetLoweringInfo();
+  return TLI.expandCTPOP(Op.getNode(), DAG);
```
TLI is this. Also I would expect this custom lowering to just be identical to what you get for expand?

andreisfr replied:
> TLI is this. Also I would expect this custom lowering to just be identical to what you get for expand?

The instruction sequence which I proposed before is a bit different from sequence generated by  TLI.expandCTPOP(). It is because I used "right shift and mask" at the last stage and thus need to use 2 EXTUI operations in my implementation. The TLI.expandCTPOP() uses "left shift and mask" at the last stage, as I understand, but instruction amount in both implementations is the same.

arsenm replied:
Also CTPOP isn't mentioned in the patch description. Should this be done separately?

andreisfr replied:
You mean to change first patch description from "[Xtensa] Implement lowering Mul/Div/Shift operations." to something like "[Xtensa] Implement lowering Mul/Div/Shift/CTTZ/CTLZ/CTPOP operations."? I can force update patches

arsenm replied:
Either that, or split the ctpop into a separate pR

s-barannikov commented on llvm/test/CodeGen/Xtensa/shift.ll:
```diff
@@ -0,0 +1,164 @@
+  ret i32 %c
+}
+
+define i32 @lshr_imm_1(i32 %x, i32 %y) nounwind {
```
(nit) %y is unused in some tests.

andreisfr replied:
Fixed. Thank you.

andreisfr committed:
[Xtensa] Fix shift tests.

andreisfr committed:
[Xtensa] Fix SHL instruction shift and add SRL test.

andreisfr committed:
[Xtensa] Minor fixes in comments.

jrtc27 commented on llvm/test/CodeGen/Xtensa/mul.ll:
```diff
@@ -0,0 +1,672 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
I assume these are being generated with your own fork, because upstream LLVM does not support xtensa in update_llc_test_checks.py:

```
Traceback (most recent call last):
  File "llvm/utils/update_llc_test_checks.py", line 303, in <module>
    main()
  File "llvm/utils/update_llc_test_checks.py", line 193, in main
    scrubber, function_re = output_type.get_run_handler(triple)
  File "llvm/utils/UpdateTestChecks/asm.py", line 592, in get_run_handler
    raise KeyError("Triple %r is not supported" % (triple))
KeyError: "Triple 'xtensa' is not supported"
```

Please upstream this, as the whole point of using UTC is that others can rerun the scripts to update the tests automatically, but they can't do that if you've got non-upstreamed changes to UTC itself.
